tools/calyx-writer/lib.rs

```diff
@@ -0,0 +1,1077 @@
+    pub fn pop(&mut self) {
+        for _ in 0..self.add_indent {
+            self.current_indent.pop();
+        }
+    }
+}
+
+impl fmt::Write for IndentFormatter<'_, '_> {
+    fn write_str(&mut self, s: &str) -> fmt::Result {
+        for c in s.chars() {
+            if self.last_was_newline {
+                self.formatter.write_str(&self.current_indent)?;
+            }
+            self.formatter.write_char(c)?;
+            self.last_was_newline = c == '\n';
+        }
+        Ok(())
+    }
+}
+
+/// A location in the source code for source element `element` and optionally a
+/// unique identifier `id` for that element. For example,
+/// ```
+/// // COMPONENT END: main
+/// ```
+/// would be a marker for the "component" element and with a unique identifier
+/// of "main".
+struct Marker {
+    element: String,
+    id: Option<String>,
+}
+
+impl Marker {
+    /// A marker for an `element`.
+    #[allow(dead_code)]
+    pub fn general<S: ToString>(element: S) -> Marker {
+        Marker {
+            element: element.to_string(),
+            id: None,
+        }
+    }
+
+    /// A marker for a unique `element` identified by `id`.
+    pub fn unique<S: ToString, T: ToString>(element: S, id: T) -> Marker {
+        Marker {
+            element: element.to_string(),
+            id: Some(id.to_string()),
+        }
+    }
+
+    /// Constructs a comment string for the marker at a given `location`. For
+    /// example, `marker.to_string("end")`.
+    pub fn to_string<S: ToString>(&self, location: S) -> String {
+        format!(
+            "// {} {}{}\n",
+            self.element.to_ascii_uppercase(),
+            location.to_string().to_ascii_uppercase(),
+            self.id
+                .as_ref()
+                .map(|id| format!(": {}", id))
+                .unwrap_or_default()
+        )
+    }
+}
+
+/// An element of calyx source code that MUST implement [`CalyxWriter::write`],
+/// can OPTIONALLY implement [`CalyxWriter::marker`], and must NEVER override
+/// [`CalyxWriter::writeln`].
+trait CalyxWriter {
+    fn marker(&self) -> Option<Marker> {
+        None
+    }
+
+    /// Writes this element to `f`. See also [`CalyxWriter::writeln`].
+    fn write(&self, f: &mut IndentFormatter<'_, '_>) -> fmt::Result;
+
+    /// Writes this element followed by a newline and wrapped with markers if
+    /// the element provides them. See also [`CalyxWriter::write`].
+    ///
+    /// Do NOT override this function. See details at [`CalyxWriter`].
+    fn writeln(&self, f: &mut IndentFormatter<'_, '_>) -> fmt::Result {
+        if let Some(marker) = self.marker() {
+            write!(f, "{}", marker.to_string("start"))?;
+            self.write(f)?;
+            f.write_char('\n')?;
+            write!(f, "{}", marker.to_string("end"))?;
+        } else {
+            self.write(f)?;
+            f.write_char('\n')?;
+        }
+        Ok(())
+    }
+}
+
+/// Imports a calyx file from the standard library.
```

Comment:
Is there a way to use `import` that doesn't search relative to the standard library? Like will it try searching relative to the current file if it can't fine it?

Reply:
I believe so? At least looking at the canonicalize fn here: https://github.com/calyxir/calyx/blob/6c2d82cd5224f32321dd0bd9b7a867e7f0e9c804/calyx-frontend/src/workspace.rs#L59

Always possible that I'm misreading though